File: sudoku-ums/ums-boot/src/main/java/com/zumin/sudoku/ums/service/StatisticsUserService.java
package com.zumin.sudoku.ums.service;


import com.zumin.sudoku.common.redis.annotation.ExtCacheable;
import com.zumin.sudoku.common.web.domain.StatisticsDateRange;
import com.zumin.sudoku.common.web.template.GetStatisticsDataTemplate;
import com.zumin.sudoku.ums.mapper.SysUserMapper;
import com.zumin.sudoku.ums.pojo.bo.StatisticsUserDataBO;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

/**
 * 统计用户信息的业务层类
 */
@Service
@RequiredArgsConstructor
public class StatisticsUserService {

  private final SysUserMapper sysUserMapper;

  /**
   * 获取在[startDate,endDate)中的用户统计信息列表
   *
   * @param dateRange 统计日期范围
   * @return 用户统计信息列表
   */
  @ExtCacheable(value = "statisticsUserData", ttlOfDays = 1)
  public List<StatisticsUserDataBO> getStatisticsUserData(StatisticsDateRange dateRange) {
    return new GetStatisticsDataTemplate<StatisticsUserDataBO>(dateRange).getData((firstDate, lastDate) -> {
      Integer newUserTotal = sysUserMapper.countNewUserByDateBetween(firstDate, lastDate);
      Integer recentLoginUserTotal = sysUserMapper.countRecentLoginUserByDateBetween(firstDate, lastDate);
      return new StatisticsUserDataBO(newUserTotal, recentLoginUserTotal);
    });
  }

  /**
   * 获取用户总数
   *
   * @return 用户总数
   */
  @ExtCacheable(value = "userTotal", ttlOfDays = 1)
  public Integer getUserTotal() {
    return sysUserMapper.selectCount(null);
  }
}

File: sudoku-ums/ums-boot/src/main/java/com/zumin/sudoku/ums/service/StatisticsUserService.kt
package com.zumin.sudoku.ums.service

import com.zumin.sudoku.common.redis.ExtCacheable
import com.zumin.sudoku.common.web.domain.StatisticsDateRange
import com.zumin.sudoku.common.web.statistics.GetStatisticsDataTemplate
import com.zumin.sudoku.ums.mapper.SysUserMapper
import com.zumin.sudoku.ums.pojo.StatisticsUserDataBO
import org.springframework.stereotype.Service

/**
 * 统计用户信息的业务层类
 */
@Service
class StatisticsUserService(
  private val sysUserMapper: SysUserMapper,
) {

  /**
   * 获取在[startDate,endDate)中的用户统计信息列表
   *
   * @param dateRange 统计日期范围
   * @return 用户统计信息列表
   */
  @ExtCacheable(value = ["statisticsUserData"], ttlOfDays = 1)
  fun getStatisticsUserData(dateRange: StatisticsDateRange): List<StatisticsUserDataBO> {
    return GetStatisticsDataTemplate<StatisticsUserDataBO>(dateRange).getData { startDate, endDate ->
      val newUserTotal = sysUserMapper.countNewUserByDateBetween(startDate, endDate)
      val recentLoginUserTotal = sysUserMapper.countRecentLoginUserByDateBetween(startDate, endDate)
      StatisticsUserDataBO(newUserTotal, recentLoginUserTotal)
    }
  }

  /**
   * 获取用户总数
   *
   * @return 用户总数
   */
  @ExtCacheable(value = ["userTotal"], ttlOfDays = 1)
  fun getUserTotal(): Int {
    return sysUserMapper.selectCount(null)
  }
}